src/IndexMarkerV2.sol

```diff
@@ -0,0 +1,298 @@
+ * trusted EOA (not a contract), but the OPENSEA_ROYALTIES_ADMIN_ROLE should
+ * be a multisig or governance contract.
+ */
+
+contract IndexMarkerV2 is
+  Initializable,
+  ERC721Upgradeable,
+  AccessControlUpgradeable,
+  EIP712Upgradeable,
+  ERC721VotesUpgradeable,
+  ERC2981Upgradeable,
+  UUPSUpgradeable,
+  IERC173 /* required for opensea royalties */
+{
+  bytes32 public constant UPGRADER_ROLE = keccak256("UPGRADER_ROLE");
+  bytes32 public constant SET_TOMB_VOTING_WEIGHT_ROLE =
+    keccak256("SET_TOMB_VOTING_WEIGHT_ROLE");
+  bytes32 public constant SET_TOMB_CONTRACTS_ROLE =
+    keccak256("SET_TOMB_CONTRACTS_ROLE");
+  bytes32 public constant SET_ROYALTIES_ROLE = keccak256("SET_ROYALTIES_ROLE");
+  bytes32 public constant UPDATE_MARKET_SETTINGS_ROLE =
+    keccak256("UPDATE_MARKET_SETTINGS_ROLE");
+  bytes32 public constant OPENSEA_ROYALTIES_ADMIN_ROLE =
+    keccak256("OPENSEA_ROYALTIES_ADMIN_ROLE");
+
+  uint16 internal constant DEFAULT_TOMB_HOLDER_VOTING_WEIGHT = 30; // 30 votes
+  uint96 internal constant DEFAULT_ROYALTY_BPS = 1_000; // 10%
+  address payable internal constant TOMB_ARTIST =
+    payable(0x4a61d76ea05A758c1db9C9b5a5ad22f445A38C46);
+
+  // zora & opensea compatibility
+  IOperatorFilterRegistry immutable operatorFilterRegistry =
+    IOperatorFilterRegistry(0x000000000000AAeB6D7670E522A718067333cd4E);
+  address public marketFilterDAOAddress;
+
+  uint96 public royaltyBps;
+  IERC721MetadataUpgradeable public indexMarker;
+  mapping(address => bool) public isTombContract;
+  mapping(address => mapping(uint256 => bool)) public isSingletonTombToken;
+  uint16 public tombHolderVotingWeight;
+  address internal openseaRoyaltiesManager;
+
+  constructor() {
+    _disableInitializers();
+  }
+
+  function initialize(
+    address _indexMarker,
+    address payable _royaltyDestination,
+    address _marketFilterDAOAddress
+  ) public initializer {
+    __ERC721_init("Wrapped Tomb Index Marker", "WMKR");
+    __AccessControl_init();
+    __EIP712_init("Wrapped Tomb Index Marker", "1");
+    __ERC721Votes_init();
+    __UUPSUpgradeable_init();
+    __ERC2981_init();
+
+    indexMarker = IERC721MetadataUpgradeable(_indexMarker);
+    marketFilterDAOAddress = _marketFilterDAOAddress;
+    tombHolderVotingWeight = DEFAULT_TOMB_HOLDER_VOTING_WEIGHT;
+    royaltyBps = DEFAULT_ROYALTY_BPS;
+
+    // uncomment below if we allow tokenId 0 to mint
+    /* isSingletonTombToken[_indexMarker][0] = true;
+    _setTokenRoyalty(0, TOMB_ARTIST, royaltyBps); */
+
+    _setDefaultRoyalty(_royaltyDestination, royaltyBps);
+    _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    openseaRoyaltiesManager = msg.sender;
+    _grantRole(OPENSEA_ROYALTIES_ADMIN_ROLE, msg.sender);
+  }
+
+  function tokenURI(uint256 tokenId)
+    public
+    view
+    override
+    returns (string memory)
+  {
+    return indexMarker.tokenURI(tokenId);
+  }
+
+  function mint(uint256 tokenId) public {
```

Comment:
Would it be possible to implement this inside of `onERC721Received` to avoid having to have users call `setApprovalForAll`?
So the flow would be people do `safeTransferFrom` on the og marker contract, and then `onERC721Received` mints the token

Reply:
yeah that should be doable!